aten/src/ATen/copy_wrapper.py

```diff
@@ -19,90 +16,199 @@
 }
 """)
 
+CUDA_INCLUDES = """\
+#undef THNN_
+#include "THC/THC.h"
+"""
+
 COPY = CodeTemplate("""\
-${THTensor}_copy${cuda}${src_scalar_name}(${state,}self_->tensor, static_cast<${src_tensor}*>(src.pImpl)->tensor);
+${THTensor}_copy${cuda}${src_scalar_name}(
+    ${state,}
+    static_cast<${dst_tensor}*>(dst.pImpl)->tensor,
+    static_cast<${src_tensor}*>(src.pImpl)->tensor);
 """)
 
 COPY_ASYNC_CPU = CodeTemplate("""\
 if (non_blocking) {
-    ${THTensor}_copyAsyncCPU(${state,}self_->tensor, static_cast<${src_tensor}*>(src.pImpl)->tensor);
+    ${THTensor}_copyAsyncCPU(
+        ${state,}
+        static_cast<${dst_tensor}*>(dst.pImpl)->tensor,
```

Comment:
I haven't looked through the rest of the code yet, so maybe I'm missing something, but did you just add the static_cast to the destinations for consistency with sources?  It so, this seems _slightly_ worse to me because if you forget to do the checked cast crazy things will happen, but in the old code it won't compile.

Reply:
Not just consistency; I'm reusing this code snippet in both the `copy` and `copy_from` bodies, which means that in the first case, the first argument is pre-casted (by the test at the prologue of the function body), but in the second case, it's the *second* argument that is pre-casted. Always applying the static cast lets me reuse the code snippet in both cases, otherwise I need two copies of the code.